### ISSUE TITLE
Use of an internal okHttp package causes issues/crashes when using a more recent version of okHttp(4+).

### ISSUE THREAD

**Issue by TWith2Sugars** [DataDog/dd-sdk-android#1282]
https://github.com/DataDog/dd-sdk-android/blob/dde54aa25954bcdcaffb45b89a84d1ee5450e7ff/dd-sdk-android/src/main/kotlin/com/datadog/android/core/internal/net/RequestUniqueIdentifier.kt#L10

According to their own docs this is a little bit [naughty](https://square.github.io/okhttp/changelogs/upgrading_to_okhttp_4/#internal-api-changes)

**Comment by 0xnm** [comment]
Hello @TWith2Sugars! Thanks for letting us know about this issue.

To better understand the impact of this, can you please share the exact error you get? It seems that the field which is used in our SDK is [still there](https://github.com/square/okhttp/blob/6b07f6227416506a44bb97012551929cd6a1dd79/okhttp/src/main/kotlin/okhttp3/internal/Util.kt#L67) for the latest version 4.10.0.

**Comment by TWith2Sugars** [comment]
Sure, FYI I'm using the `DatadogInterceptor`, `TracingInterceptor`, and `DatadogEventListener.Factory` as part of the `OkHttpClient.Builder`

but here is the exception:

```
E/AndroidRuntime: FATAL EXCEPTION: OkHttp Dispatcher
    Process: REDACTED, PID: 2258
    java.lang.NoClassDefFoundError: Failed resolution of: Lokhttp3/internal/Util;
        at com.datadog.android.core.internal.net.RequestUniqueIdentifierKt.identifyRequest(RequestUniqueIdentifier.kt:19)
        at com.datadog.android.DatadogEventListener$Factory.create(DatadogEventListener.kt:231)
        at okhttp3.internal.connection.RealCall.<init>(RealCall.kt:70)
        at okhttp3.OkHttpClient.newCall(OkHttpClient.kt:281)

Caused by: java.lang.ClassNotFoundException: Didn't find class "okhttp3.internal.Util" on path: DexPathList[[zip file REDACTED ]]
        at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:259)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:379)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:312)
        at com.datadog.android.core.internal.net.RequestUniqueIdentifierKt.identifyRequest(RequestUniqueIdentifier.kt:19) 
        at com.datadog.android.DatadogEventListener$Factory.create(DatadogEventListener.kt:231) 
```

I've removed some sensitive parts.

Version of okHttp we use is: **4.9.3**

**Comment by 0xnm** [comment]
Hello @TWith2Sugars. Thanks for the info!

I cannot reproduce this issue with our sample app: I specified OkHttp version `4.9.3` explicitly (below you can find the output of the `dependencies` command to see the version resolved), but don't have any crash (I tried also with code obfuscation enabled - same result).

```
+--- project :dd-sdk-android
...
|    +--- com.squareup.okhttp3:okhttp:3.12.13 -> 4.9.3
```

Can you please give additional details:

* Can you please check that OkHttp version resolved for the runtime classpath is indeed `4.9.3` by running `dependencies` task and supplying the necessary configuration name argument?
* Do you use `multiDexEnabled = true`?
* Does the stacktrace you provided have the final cause? Maybe class is not loaded, because there is some other cause down the stacktrace.
* What is the Datadog SDK version you are using?

While it is no doubt that we should move away from `okhttp3.internal` package usage, I would like to understand the urgency of this change, because the class mentioned (and property referenced) should still be there for the latest OkHttp 4.x version.

**Comment by TWith2Sugars** [comment]
Hey @0xnm,

Thanks for looking in to this btw.

- Yes I can confirm the version is `4.9.3` using the `dependencies` task againsst the `RuntimeClasspath`
- Yes `multiDexEnabled` is enabled
- Stacktrace does not have any other cause apart from the one I listed above.
- dd version is `1.17.0`

**Comment by 0xnm** [comment]
Thanks for the quick reply @TWith2Sugars! This issue sounds strange then.

Can I please ask you to do one more check? Can you please check if the final APK contains `okhttp3.internal.Util` in dex file(s)? You can use `Build` -> `Analyze APK` for that, for example.

Here is a screenshot as the reference for our sample app:

![image](https://user-images.githubusercontent.com/4046447/219042398-2392c780-7f03-47e6-b0f8-0d91c9eb1d8e.png)

**Comment by TWith2Sugars** [comment]
This is what I have:

![image](https://user-images.githubusercontent.com/234112/219043347-de14ced8-0df5-470a-829b-1e8aa3788c4b.png)

**Comment by 0xnm** [comment]
@TWith2Sugars You need to check it in the content of one of the `dex` files you have, this is just `okhttp3` meta on your screenshot, not executable file content.

**Comment by TWith2Sugars** [comment]
Ah sorry,

Turns out I do have it‽

![image](https://user-images.githubusercontent.com/234112/219045303-a0a0d35e-e28d-4fb6-9a0b-4e0a7c4539c8.png)

**Comment by 0xnm** [comment]
Yes, it means you have it in the runtime classpath (which is expected, because as I said `4.9.3` has still both this class and property). At this point it seems like some issue with class loading for your app.

I can only assume that maybe Datadog classes and OkHttp classes are in the different `dex` files and this is causing the issue, but then root cause lies on the AGP side I believe.

It means that even if we remove the usage `okhttp3.internal` package (and for sure we will later), there is a chance that similar issue pop up for something else.

**Comment by TWith2Sugars** [comment]
Thanks for the help, this at least narrows down my focus.

Much appreciated btw.

Feel free to close this issue then if you think it no longer applies.

**Comment by 0xnm** [comment]
I will keep it open to track `okhttp3.internal` package usage removal. Thanks for reporting it and for the collaboration!

**Comment by TWith2Sugars** [comment]
Just to update this for anyone else caught in the same situation.

Turns out another package has a transitive dep on okhttp 5+ which is causing the issue I'm having.

**Comment by 0xnm** [comment]
Hello @TWith2Sugars. We removed the usage of `okhttp3.internal` package, the change will be available in `1.18.0` or `1.17.1` (whatever comes first).

**Comment by 0xnm** [comment]
We have released version [1.17.1](https://github.com/DataDog/dd-sdk-android/releases/tag/1.17.1) which includes the necessary fix.
